File: presentation/src/main/java/info/firozansari/presentation/view/adapter/UsersAdapter.java

package info.firozansari.presentation.view.adapter;

import android.content.Context;

import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import info.firozansari.presentation.R;
import info.firozansari.presentation.model.UserModel;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import javax.inject.Inject;

/**
 * Adaptar that manages a collection of {@link UserModel}.
 */
public class UsersAdapter extends RecyclerView.Adapter<UsersAdapter.UserViewHolder> {

  public interface OnItemClickListener {
    void onUserItemClicked(UserModel userModel);
  }

  private List<UserModel> usersCollection;
  private final LayoutInflater layoutInflater;

  private OnItemClickListener onItemClickListener;

  @Inject
  UsersAdapter(Context context) {
    this.layoutInflater =
        (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    this.usersCollection = Collections.emptyList();
  }

  @Override public int getItemCount() {
    return (this.usersCollection != null) ? this.usersCollection.size() : 0;
  }

  @Override public UserViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    final View view = this.layoutInflater.inflate(R.layout.row_user, parent, false);
    return new UserViewHolder(view);
  }

  @Override public void onBindViewHolder(UserViewHolder holder, final int position) {
    final UserModel userModel = this.usersCollection.get(position);
    holder.textViewTitle.setText(userModel.getFullName());
    holder.itemView.setOnClickListener(new View.OnClickListener() {
      @Override public void onClick(View v) {
        if (UsersAdapter.this.onItemClickListener != null) {
          UsersAdapter.this.onItemClickListener.onUserItemClicked(userModel);
        }
      }
    });
  }

  @Override public long getItemId(int position) {
    return position;
  }

  public void setUsersCollection(Collection<UserModel> usersCollection) {
    this.validateUsersCollection(usersCollection);
    this.usersCollection = (List<UserModel>) usersCollection;
    this.notifyDataSetChanged();
  }

  public void setOnItemClickListener (OnItemClickListener onItemClickListener) {
    this.onItemClickListener = onItemClickListener;
  }

  private void validateUsersCollection(Collection<UserModel> usersCollection) {
    if (usersCollection == null) {
      throw new IllegalArgumentException("The list cannot be null");
    }
  }

  static class UserViewHolder extends RecyclerView.ViewHolder {
    @BindView(R.id.title) TextView textViewTitle;

    UserViewHolder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);
    }
  }
}

File: presentation/src/main/java/info/firozansari/presentation/view/adapter/UsersAdapter.kt
package info.firozansari.presentation.view.adapter

import android.content.Context
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView

/**
 * Adaptar that manages a collection of [UserModel].
 */
class UsersAdapter @Inject internal constructor(context: Context?) :
    RecyclerView.Adapter<UserViewHolder?>() {
    interface OnItemClickListener {
        open fun onUserItemClicked(userModel: UserModel?)
    }

    private var usersCollection: MutableList<UserModel?>?
    private val layoutInflater: LayoutInflater?
    private var onItemClickListener: OnItemClickListener? = null
    override fun getItemCount(): Int {
        return if (usersCollection != null) usersCollection.size else 0
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): UserViewHolder? {
        val view: View = layoutInflater.inflate(R.layout.row_user, parent, false)
        return UserViewHolder(view)
    }

    override fun onBindViewHolder(holder: UserViewHolder?, position: Int) {
        val userModel: UserModel? = usersCollection.get(position)
        holder.textViewTitle.setText(userModel.getFullName())
        holder.itemView.setOnClickListener {
            if (onItemClickListener != null) {
                onItemClickListener.onUserItemClicked(userModel)
            }
        }
    }

    override fun getItemId(position: Int): Long {
        return position
    }

    fun setUsersCollection(usersCollection: MutableCollection<UserModel?>?) {
        validateUsersCollection(usersCollection)
        this.usersCollection = usersCollection as MutableList<UserModel?>?
        notifyDataSetChanged()
    }

    fun setOnItemClickListener(onItemClickListener: OnItemClickListener?) {
        this.onItemClickListener = onItemClickListener
    }

    private fun validateUsersCollection(usersCollection: MutableCollection<UserModel?>?) {
        requireNotNull(usersCollection) { "The list cannot be null" }
    }

    internal class UserViewHolder(itemView: View?) : RecyclerView.ViewHolder(itemView) {
        @BindView(R.id.title)
        var textViewTitle: TextView? = null

        init {
            ButterKnife.bind(this, itemView)
        }
    }

    init {
        layoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        usersCollection = emptyList()
    }
}